codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -158,7 +158,10 @@ class KotlinJmespathExpressionVisitor(
                 append("if ($isNullExpr) null else ")
             }
 
-            append("${left.identifier}.compareTo(${right.identifier}) ${expression.comparator} 0")
+            val unSafeComparatorExpr = "compareTo(${right.identifier}) ${expression.comparator} 0"
+            val comparatorExpr = if (left.nullable) buildString { append("?.$unSafeComparatorExpr") } else buildString { append(".$unSafeComparatorExpr") }
+
+            append("${left.identifier}$comparatorExpr")
```

Comment:
**Nit**: `buildString { append(foo) }` can be rewritten as `foo`:

```kotlin
val comparatorExpr = if (left.nullable) "?.$unSafeComparatorExpr" else ".$unSafeComparatorExpr"
```

`buildString` is still a nice function and may be useful here if extended around the wider logic:

```kotlin
val comparison = buildString {
    append(left.identifier)
    if (left.nullable) append('?')
    append(".compareTo(${right.identifier}) ${expression.comparator} 0")
}
append(comparison)
```